include/ginkgo/extensions/kokkos/spaces.hpp

```diff
@@ -0,0 +1,280 @@
+ * Throws if the memory space is ~not~ accessible by the executor associated
+ * with the passed in Ginkgo object.
+ *
+ * @tparam MemorySpace  A Kokkos memory space type..
+ * @tparam T  A Ginkgo type that has the member function `get_executor`.
+ *
+ * @param obj  Object which executor is used to check the access to the memory
+ * space.
+ */
+template <typename MemorySpace, typename T>
+inline void assert_compatibility(T&& obj)
+{
+    GKO_THROW_IF_INVALID(check_compatibility<MemorySpace>(obj.get_executor()),
+                         "Executor type and memory space are incompatible");
+}
+
+
+}  // namespace detail
+
+
+/**
+ * Creates an Executor matching the Kokkos::DefaultHostExecutionSpace.
+ *
+ * If no kokkos host execution space is enabled, this throws an exception.
+ *
+ * @return  An executor of type either ReferenceExecutor or OmpExecutor.
+ */
+inline std::shared_ptr<Executor> create_default_host_executor()
+{
+#ifdef KOKKOS_ENABLE_SERIAL
+    if constexpr (std::is_same_v<Kokkos::DefaultHostExecutionSpace,
+                                 Kokkos::Serial>) {
+        return ReferenceExecutor::create();
+    }
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+    if constexpr (std::is_same_v<Kokkos::DefaultHostExecutionSpace,
+                                 Kokkos::OpenMP>) {
+        return OmpExecutor::create();
+    }
+#endif
+    GKO_NOT_IMPLEMENTED;
+}
+
+
+/**
+ * Creates an Executor for a specific Kokkos ExecutionSpace.
+ *
+ * This function supports the following Kokkos ExecutionSpaces:
+ * - Serial
+ * - OpenMP
+ * - Cuda
+ * - HIP
+ * - Experimental::SYCL
+ * If none of these spaces are enabled, then this function throws an exception.
+ * For Cuda, HIP, SYCL, the device-id used by Kokkos is passed to the Executor
+ * constructor.
+ *
+ * @tparam ExecSpace  A supported Kokkos ExecutionSpace.
+ * @tparam MemorySpace  A supported Kokkos MemorySpace. Defaults to the one
+ *                      defined in the ExecSpace.
+ *
+ * @param ex  the execution space
+ *
+ * @return  An executor matching the type of the ExecSpace.
+ */
+template <typename ExecSpace,
+          typename MemorySpace = typename ExecSpace::memory_space>
+inline std::shared_ptr<Executor> create_executor(ExecSpace ex, MemorySpace = {})
+{
+    static_assert(
+        Kokkos::SpaceAccessibility<ExecSpace, MemorySpace>::accessible);
+#ifdef KOKKOS_ENABLE_SERIAL
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::Serial>) {
+        return ReferenceExecutor::create();
+    }
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::OpenMP>) {
+        return OmpExecutor::create();
+    }
+#endif
+#ifdef KOKKOS_ENABLE_CUDA
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::Cuda>) {
+        if constexpr (std::is_same_v<MemorySpace, Kokkos::CudaSpace>) {
+            return CudaExecutor::create(
+                Kokkos::device_id(), create_default_host_executor(),
+                std::make_shared<CudaAllocator>(), ex.cuda_stream());
```

Comment:
Is it safe to assume that `cuda_stream` is the only stream associated with the execution space ? Same question for the other execution spaces.

Reply:
Kokkos API doesn't offer any other stream access, so I would assume so. The stream depends on the execution space object, so you could have multiple execution spaces with different streams. But that should be covered here.